datasource/timeouts/timeouts.go

```diff
@@ -0,0 +1,109 @@
+package timeouts
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+const defaultTimeout = time.Minute * 20
+
+// Type is an attribute type that represents timeouts.
+type Type struct {
+	types.ObjectType
+}
+
+// ValueFromTerraform returns a Value given a tftypes.Value.
+// Value embeds the types.Object value returned from calling ValueFromTerraform on the
+// types.ObjectType embedded in Type.
+func (t Type) ValueFromTerraform(ctx context.Context, in tftypes.Value) (attr.Value, error) {
+	val, err := t.ObjectType.ValueFromTerraform(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, ok := val.(types.Object)
+	if !ok {
+		return nil, fmt.Errorf("%T cannot be used as types.Object", val)
+	}
+
+	return Value{
+		obj,
+	}, err
+}
+
+// Equal returns true if `candidate` is also an Type and has the same
+// AttributeTypes.
+func (t Type) Equal(candidate attr.Type) bool {
+	other, ok := candidate.(Type)
+	if !ok {
+		return false
+	}
+
+	return t.ObjectType.Equal(other.ObjectType)
+}
+
+// Value represents an object containing values to be used as time.Duration for timeouts.
+type Value struct {
+	types.Object
+}
+
+// Equal returns true if the Value is considered semantically equal
+// (same type and same value) to the attr.Value passed as an argument.
+func (t Value) Equal(c attr.Value) bool {
+	other, ok := c.(Value)
+
+	if !ok {
+		return false
+	}
+
+	return t.Object.Equal(other.Object)
+}
+
+// Type returns an Type with the same attribute types as `t`.
+func (t Value) Type(ctx context.Context) attr.Type {
+	return Type{
+		types.ObjectType{
+			AttrTypes: t.AttributeTypes(ctx),
+		},
+	}
+}
+
+// Read attempts to retrieve the "read" attribute and parse it as time.Duration.
+// If any diagnostics are generated they are returned along with the default timeout of 20 minutes.
+func (t Value) Read(ctx context.Context) (time.Duration, diag.Diagnostics) {
+	return t.getTimeout(ctx, attributeNameRead)
+}
+
+func (t Value) getTimeout(_ context.Context, timeoutName string) (time.Duration, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	value, ok := t.Object.Attributes()[timeoutName]
+	if !ok {
+		diags.Append(diag.NewErrorDiagnostic(
+			"Timeout Does Not Exist",
+			fmt.Sprintf("timeout for %q does not exist", timeoutName),
+		))
+
+		return defaultTimeout, diags
+	}
```

Comment:
How do you feel about the prior implementation that accepted a default timeout duration inline? I think returning an error diagnostic here could be problematic for provider implementations which elsewhere in the framework do not need to conditionally choose when to append diagnostics to their response diagnostics. I'd expect something like the following to "always work" (configured or not):

```go
timeout, diags := data.Timeouts.Read(ctx) // or Read(ctx, 20*time.Minute)

resp.Diagnostics(diags...)
```

If we wanted to make this nicer for troubleshooting, I'd suggest passing `context.Context` through and logging using `tflog.Info(ctx, timeoutName + " timeout configuration not found, using provided default")`

Reply:
I think the super cool thing here is that once we have a custom type for "Go time durations", we can likely remove the `diag.Diagnostics` return completely. 😄